Add kvikio blogpost (#308)
xarray-contrib/xarray.dev: src/posts/xarray-kvikio/index.md

---
title: "Enabling GPU-native analytics with Xarray and Kvikio"
date: "2022-06-09"
authors:
  - name: Deepak Cherian
    github: dcherian
summary: "An experiment with direct-to-GPU reads from a Zarr store using Xarray."
---

## TLDR

We [demonstrate](https://github.com/xarray-contrib/cupy-xarray/pull/10) registering an Xarray backend that reads data from a Zarr store directly to GPU memory as [CuPy arrays](https://cupy.dev) using the new [Kvikio library](https://docs.rapids.ai/api/kvikio/stable/) and [GPU Direct Storage](https://developer.nvidia.com/blog/gpudirect-storage/) technology.

## Background

### What is GPU Direct Storage

> Insert https://developer.nvidia.com/blog/wp-content/uploads/2019/08/GPUDirect-Fig-1-New.png somehow

Author: Need to add text + image here.

```python
import xarray as xr

ds = xr.open_dataset("file.zarr", engine="kvikio", consolidated=False)
```

With this `ds.load()` will load directly to GPU memory and `ds` will now contain CuPy arrays.

At present there are a few limitations:

1. stores cannot be read with consolidated metadata, and
2. compression is unsupported by the backend.

## Quick demo

First create an example uncompressed dataset to read from

```
import xarray as xr

store = "./air-temperature.zarr"

airt = xr.tutorial.open_dataset("air_temperature", engine="netcdf4")

for var in airt.variables:
    airt[var].encoding["compressor"] = None
airt.to_zarr(store, mode="w", consolidated=True)
```

Now read

```
# consolidated must be False
ds = xr.open_dataset(store, engine="kvikio", consolidated=False)
ds
```

Now load a small subset

```python
type(ds["air"].isel(time=0, lat=10).load().data)
```

```
cupy._core.core.ndarray
```

Success!

Xarray integrates [decently well](https://cupy-xarray.readthedocs.io/quickstart.html) with CuPy arrays so you should be able to test out analysis pipelines pretty seamlessly.

## Cool demo

> We don't have a cool demo yet but are looking to develop one very soon!

Author: Link to https://github.com/pangeo-data/xbatcher/issues/87?

Reviewer: Sure, and/or we can link to the forum at https://discourse.pangeo.io/tag/machine-learning if people want to reach out.

---
title: "Enabling GPU-native analytics with Xarray and kvikIO"
date: "2022-08-25"
authors:

>   - name: Deepak Cherian, Wei Ji Leong
>     github: dcherian, weiji14

Reviewer 2: ```suggestion
  - name: Deepak Cherian
    github: dcherian
  - name: Wei Ji Leong
    github: weiji14
```

summary: "An experiment with direct-to-GPU reads from a Zarr store using Xarray."
---

## TLDR

We [demonstrate](https://github.com/xarray-contrib/cupy-xarray/pull/10) registering an Xarray backend that reads data from a Zarr store directly to GPU memory as [CuPy arrays](https://cupy.dev) using the new [kvikIO library](https://docs.rapids.ai/api/kvikio/stable/) and [GPU Direct Storage](https://developer.nvidia.com/blog/gpudirect-storage/) technology. This allows direct-to-GPU reads and GPU-native analytics on existing pipelines 🎉 😱 🤯 🥳.

## Background

### What is GPU Direct Storage?

Quoting [this nVIDIA blogpost](https://developer.nvidia.com/blog/gpudirect-storage/)

> > I/O, the process of loading data from storage to GPUs for processing, has historically been controlled by the CPU. As computation shifts from slower CPUs to faster GPUs, I/O becomes more of a bottleneck to overall application performance. Just as GPUDirect RDMA (Remote Direct Memory Address) improved bandwidth and latency when moving data directly between a network interface card (NIC) and GPU memory, a new technology called GPUDirect Storage enables a direct data path between local or remote storage, like NVMe or NVMe over Fabric (NVMe-oF), and GPU memory. Both GPUDirect RDMA and GPUDirect Storage avoid extra copies through a bounce buffer in the CPU’s memory and enable a direct memory access (DMA) engine near the NIC or storage to move data on a direct path into or out of GPU memory, all without burdening the CPU or GPU. For GPUDirect Storage, storage location doesn’t matter; it could be inside an enclosure, within the rack, or connected over the network.

Author: @weiji14 this didn;'t reneder correctly with newlines.

Also thanks for the repeated reviews, I added you as co-author (I hope the syntax works)

---
title: 'Enabling GPU-native analytics with Xarray and kvikIO'

> date: '2022-08-25'

Reviewer: Need to update this date before publishing.

Do we want to publish the blog post soon-ish (e.g. by end of August) or wait for a bit longer (e.g. until some of those WIP PRs get a bit more stable)? Btw, I don't have merge permissions on this repo...

Reviewer 2: @weiji14, i updated the date. Regarding whether we should wait or ship this as is, i'll defer to you :)

Reviewer: Let's ship it :rocket:

For Xarray, the key bit is that kvikIO exposes a zarr store [`kvikio.zarr.GDSStore`](https://docs.rapids.ai/api/kvikio/stable/api.html#zarr) that does all the hard work for us. Since Xarray knows how to read Zarr stores, we can adapt it to create a new storage backend that uses `kvikio`. And thanks to recent work funded by the [Chan Zuckerberg Initiative](https://xarray.dev/blog/czi-eoss-grant-conclusion), creating and registering a [new backend](https://docs.xarray.dev/en/stable/internals/how-to-add-new-backend.html) is quite easy!

## Integrating with Xarray

Getting all this to work nicely requires using three in-progress pull requests that

1. [Teach Zarr to handle alternative array classes](https://github.com/zarr-developers/zarr-python/pull/934)
2. [Rewrite a small bit of Xarray to not cast all data to a numpy array after read from disk](https://github.com/pydata/xarray/pull/6874)
3. [Make a backend that connects Xarray to kvikIO](https://github.com/xarray-contrib/cupy-xarray/pull/10)

> Writing the backend for Xarray was relatively easily. Most of the code was copied over from the existing Zarr backend. Most of the effort was in ensuring that dimension coordinates could be read in directly to host memory without raising an error. This is required because Xarrays creates `pandas.Index` objects for such variables. In the future, we could consider using `cudf.Index` instead to allow a fully GPU-backed Xarray object.

Author: ```suggestion
Writing the backend for Xarray was relatively easy with most of the code copied over or inherited from the existing Zarr backend. We did have to ensure that dimension coordinates could be read in directly to host memory without raising an error (by default kvikIO loads all data to device memory). This is required because Xarrays creates `pandas.Index` objects for such variables. In the future, we could consider using `cudf.Index` instead to allow a fully GPU-backed Xarray object.
```

## Usage

> Assuming you have all the pieces together (see [Appendix I]() and [Appendix II]() for step-by-step instructions), then using all this cool technology only requires adding `engine="kvikio"` to your `open_dataset` line (!)

Author: Whats the syntax for inline links to Appendix I and II below?

> Assuming you have all the pieces together (see [Appendix I]() and [Appendix II]() for step-by-step instructions), then using all this cool technology only requires adding `engine="kvikio"` to your `open_dataset` line (!)

Reviewer 2: ```suggestion
Assuming you have all the pieces together (see [Appendix I](#appendix-i--step-by-step-install-instructions) and [Appendix II](#appendix-ii--making-sure-gds-is-working) for step-by-step instructions), then using all this cool technology only requires adding `engine="kvikio"` to your `open_dataset` line (!)
```

```python
import xarray as xr

ds = xr.open_dataset("file.zarr", engine="kvikio", consolidated=False)
```

Notice that importing `cupy_xarray` was not needed. `cupy_xarray` uses [entrypoints](https://packaging.python.org/en/latest/specifications/entry-points/) to register the kvikIO backend with Xarray.

With this `ds.load()` will load directly to GPU memory and `ds` will now contain CuPy arrays. At present there are a few limitations:

1. Zarr stores cannot be read with consolidated metadata, and
2. compression is unsupported by the kvikIO backend.

## Quick demo

First create an example uncompressed dataset to read from

```
import xarray as xr

store = "./air-temperature.zarr"

airt = xr.tutorial.open_dataset("air_temperature", engine="netcdf4")

for var in airt.variables:
    airt[var].encoding["compressor"] = None
airt.to_zarr(store, mode="w", consolidated=True)
```

Now read

```
# consolidated must be False
ds = xr.open_dataset(store, engine="kvikio", consolidated=False)
ds.air
```

> ```
> <xarray.DataArray 'air' (time: 2920, lat: 25, lon: 53)>

Reviewer: Not sure if syntax highlighting helps here, but won't hurt I guess.
```suggestion
```python
<xarray.DataArray 'air' (time: 2920, lat: 25, lon: 53)>
```

  * lon      (lon) float32 200.0 202.5 205.0 207.5 ... 322.5 325.0 327.5 330.0
  * time     (time) datetime64[ns] 2013-01-01 ... 2014-12-31T18:00:00
Attributes:
    GRIB_id:       11
    GRIB_name:     TMP
    actual_range:  [185.16000366210938, 322.1000061035156]
    dataset:       NMC Reanalysis
    level_desc:    Surface
    long_name:     4xDaily Air temperature at sigma level 995
    parent_stat:   Other
    precision:     2
    statistic:     Individual Obs
    units:         degK
    var_desc:      Air temperature
```

Note that we get Xarray's lazy backend arrays by default, and that dimension coordinate variables `lat`, `lon`, `time` were read. At this point this looks identical to what we get with a standard `xr.open_dataset(store, engine="zarr")` command.

Now load a small subset

```python
type(ds["air"].isel(time=0, lat=10).load().data)
```

```
cupy._core.core.ndarray
```

Success!

Xarray integrates [decently well](https://cupy-xarray.readthedocs.io/quickstart.html) with CuPy arrays so you should be able to test out analysis pipelines pretty easily.

## Cool demo

We don't have a cool demo yet but are looking to develop one very soon!

> Reach out if you have ideas. We would love to hear from you.

Author: ```suggestion
[Reach out](https://discourse.pangeo.io/tag/machine-learning) if you have ideas. We would love to hear from you.
```

## Summary

We demonstrate integrating the kvikIO library using Xarray's new backend entrypoints. With everything set up, simply adding `engine="kvikio"` enables direct-to-GPU reads from disk or over the network.

## Acknowledgments

My time on this project was funded by NASA-OSTFL 80NSSC22K0345 "Enhancing analysis of NASA data with the open-source Python Xarray Library"

## Appendix I : Step-by-step install instructions

[Wei Ji Leong](https://github.com/weiji14) helpfully [provided steps](https://github.com/xarray-contrib/cupy-xarray/pull/10#issuecomment-1218374773) to get started on your machine:

> ```

Reviewer: See if there's syntax highlighting for bash

```suggestion
```bash
```

    units:         degK
    var_desc:      Air temperature
```

Note that we get Xarray's lazy backend arrays by default, and that dimension coordinate variables `lat`, `lon`, `time` were read. At this point this looks identical to what we get with a standard `xr.open_dataset(store, engine="zarr")` command.

Now load a small subset

```python
type(ds["air"].isel(time=0, lat=10).load().data)
```

```
cupy._core.core.ndarray
```

Success!

Xarray integrates [decently well](https://cupy-xarray.readthedocs.io/quickstart.html) with CuPy arrays so you should be able to test out analysis pipelines pretty easily.

## Cool demo

We don't have a cool demo yet but are looking to develop one very soon!

Reach out if you have ideas. We would love to hear from you.

## Summary

We demonstrate integrating the kvikIO library using Xarray's new backend entrypoints. With everything set up, simply adding `engine="kvikio"` enables direct-to-GPU reads from disk or over the network.

## Acknowledgments

My time on this project was funded by NASA-OSTFL 80NSSC22K0345 "Enhancing analysis of NASA data with the open-source Python Xarray Library"

## Appendix I : Step-by-step install instructions

> [Wei Ji Leong](https://github.com/weiji14) helpfully [provided steps](https://discourse.pangeo.io/t/favorite-way-to-go-from-netcdf-xarray-to-torch-tf-jax-et-al/2663/2) to get started on your machine:

Reviewer: Point to correct link at https://github.com/xarray-contrib/cupy-xarray/pull/10#issuecomment-1218374773
```suggestion
[Wei Ji Leong](https://github.com/weiji14) helpfully [provided steps](https://github.com/xarray-contrib/cupy-xarray/pull/10#issuecomment-1218374773) to get started on your machine:
```
